# Comparing `tmp/genofunc-1.1.1-py3-none-any.whl.zip` & `tmp/genofunc-1.1.2-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,36 +1,36 @@
-Zip file size: 28765 bytes, number of entries: 34
+Zip file size: 28965 bytes, number of entries: 34
 -rw-r--r--  2.0 unx      537 b- defN 22-Mar-25 16:35 genofunc/__init__.py
--rw-r--r--  2.0 unx    16448 b- defN 23-May-10 09:17 genofunc/__main__.py
+-rw-r--r--  2.0 unx    16649 b- defN 23-May-18 10:01 genofunc/__main__.py
 -rw-r--r--  2.0 unx     1419 b- defN 22-Mar-25 16:35 genofunc/concatenate_fasta.py
 -rw-r--r--  2.0 unx     2932 b- defN 22-Mar-25 16:35 genofunc/extract_metadata.py
 -rw-r--r--  2.0 unx     3781 b- defN 23-May-11 11:44 genofunc/feature_extractor.py
 -rw-r--r--  2.0 unx     4847 b- defN 22-Mar-25 16:35 genofunc/filter_fasta.py
 -rw-r--r--  2.0 unx     1576 b- defN 22-Mar-25 16:35 genofunc/gene_concatenator.py
--rw-r--r--  2.0 unx     6117 b- defN 23-May-10 09:16 genofunc/genome_annotator.py
+-rw-r--r--  2.0 unx     6411 b- defN 23-May-19 11:58 genofunc/genome_annotator.py
 -rw-r--r--  2.0 unx     7691 b- defN 22-Jun-10 10:33 genofunc/genome_annotator2.py
 -rw-r--r--  2.0 unx     4677 b- defN 22-Mar-25 16:35 genofunc/group_align.py
 -rw-r--r--  2.0 unx     4106 b- defN 22-Mar-25 16:35 genofunc/merge.py
 -rw-r--r--  2.0 unx     1587 b- defN 22-Mar-25 16:35 genofunc/name_splitter.py
 -rw-r--r--  2.0 unx     3924 b- defN 23-May-15 10:21 genofunc/reference_matcher.py
 -rw-r--r--  2.0 unx     1521 b- defN 22-Mar-25 16:35 genofunc/rename_fasta.py
 -rw-r--r--  2.0 unx     2789 b- defN 22-Mar-25 16:35 genofunc/strain_encoder.py
 -rw-r--r--  2.0 unx      650 b- defN 22-Mar-25 16:35 genofunc/utils.py
 -rw-r--r--  2.0 unx      410 b- defN 22-Mar-25 16:35 genofunc/subcommands/__init__.py
 -rw-r--r--  2.0 unx      350 b- defN 22-Mar-25 16:35 genofunc/subcommands/concatenate_fasta.py
 -rw-r--r--  2.0 unx      536 b- defN 22-Mar-25 16:35 genofunc/subcommands/extract_metadata.py
 -rw-r--r--  2.0 unx      418 b- defN 22-Mar-25 16:35 genofunc/subcommands/feature_extractor.py
 -rw-r--r--  2.0 unx      448 b- defN 22-Mar-25 16:35 genofunc/subcommands/filter_fasta.py
 -rw-r--r--  2.0 unx      308 b- defN 22-Mar-25 16:35 genofunc/subcommands/gene_concatenator.py
--rw-r--r--  2.0 unx      334 b- defN 22-Mar-25 16:35 genofunc/subcommands/genome_annotator.py
+-rw-r--r--  2.0 unx      359 b- defN 23-May-18 10:02 genofunc/subcommands/genome_annotator.py
 -rw-r--r--  2.0 unx      342 b- defN 22-Mar-25 16:35 genofunc/subcommands/group_align.py
 -rw-r--r--  2.0 unx      325 b- defN 22-Mar-25 16:35 genofunc/subcommands/merge.py
 -rw-r--r--  2.0 unx      373 b- defN 22-Mar-25 16:35 genofunc/subcommands/name_splitter.py
 -rw-r--r--  2.0 unx      391 b- defN 23-May-10 09:17 genofunc/subcommands/reference_matcher.py
 -rw-r--r--  2.0 unx      317 b- defN 22-Mar-25 16:35 genofunc/subcommands/rename_fasta.py
 -rw-r--r--  2.0 unx      397 b- defN 22-Mar-25 16:35 genofunc/subcommands/strain_encoder.py
--rw-r--r--  2.0 unx      572 b- defN 23-May-15 10:28 genofunc-1.1.1.dist-info/METADATA
--rw-r--r--  2.0 unx       92 b- defN 23-May-15 10:28 genofunc-1.1.1.dist-info/WHEEL
--rw-r--r--  2.0 unx       53 b- defN 23-May-15 10:28 genofunc-1.1.1.dist-info/entry_points.txt
--rw-r--r--  2.0 unx        9 b- defN 23-May-15 10:28 genofunc-1.1.1.dist-info/top_level.txt
-?rw-rw-r--  2.0 unx     2919 b- defN 23-May-15 10:28 genofunc-1.1.1.dist-info/RECORD
-34 files, 73196 bytes uncompressed, 24045 bytes compressed:  67.1%
+-rw-r--r--  2.0 unx      572 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/METADATA
+-rw-r--r--  2.0 unx       92 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/WHEEL
+-rw-r--r--  2.0 unx       53 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/entry_points.txt
+-rw-r--r--  2.0 unx        9 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/top_level.txt
+?rw-rw-r--  2.0 unx     2919 b- defN 23-May-19 12:03 genofunc-1.1.2.dist-info/RECORD
+34 files, 73716 bytes uncompressed, 24245 bytes compressed:  67.1%
```

## zipnote {}

```diff
@@ -81,23 +81,23 @@
 
 Filename: genofunc/subcommands/rename_fasta.py
 Comment: 
 
 Filename: genofunc/subcommands/strain_encoder.py
 Comment: 
 
-Filename: genofunc-1.1.1.dist-info/METADATA
+Filename: genofunc-1.1.2.dist-info/METADATA
 Comment: 
 
-Filename: genofunc-1.1.1.dist-info/WHEEL
+Filename: genofunc-1.1.2.dist-info/WHEEL
 Comment: 
 
-Filename: genofunc-1.1.1.dist-info/entry_points.txt
+Filename: genofunc-1.1.2.dist-info/entry_points.txt
 Comment: 
 
-Filename: genofunc-1.1.1.dist-info/top_level.txt
+Filename: genofunc-1.1.2.dist-info/top_level.txt
 Comment: 
 
-Filename: genofunc-1.1.1.dist-info/RECORD
+Filename: genofunc-1.1.2.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## genofunc/__main__.py

```diff
@@ -1,14 +1,13 @@
 """
 This file is part of genofunc (https://github.com/xiaoyu518/genofunc).
 Copyright 2022 Xiaoyu Yu (xiaoyu.yu@ed.ac.uk) & Rachel Colquhoun (rachel.colquhoun@ed.ac.uk).
 """
 
 import argparse
-
 import genofunc
 import genofunc.subcommands
 
 def main(args=None):
     parser = argparse.ArgumentParser(
         prog="genofunc",
         description="Miscellaneous genome manipulation tools",
@@ -194,14 +193,18 @@
         help='One or more CSV or TSV tables of metadata'
     )
     subparser_genome_annotator.add_argument(
         '--reference-sequence', dest='reference_sequence', metavar='<filename>', required=True,
         help='Annotated reference sequences in json format'
     )
     subparser_genome_annotator.add_argument(
+        '--threads', dest='threads', metavar='<integer>', required=False, default=1,
+        help='Number of threads for multiprocessing (Default: 1)'
+    )
+    subparser_genome_annotator.add_argument(
         '--out-annotation', dest='out_annotation', metavar='<filename>', required=False, default="referenced.json",
         help='Output list of sequences annotated in json format'
     )
 
     subparser_genome_annotator.set_defaults(func=genofunc.subcommands.genome_annotator.run)
 
     # _______________________________  group_align  __________________________________#
```

## genofunc/genome_annotator.py

```diff
@@ -1,160 +1,180 @@
 #!/usr/bin/env python3
 
 """
 Name: genome_annotator.py
 Author: Xiaoyu Yu
-Date: 10 April 2023
+Date: 18 May 2023
 Description: Annotate genomes based on closest reference sequence annotation. 
 
 Options:
     :param raw_fasta: Raw sequences with reference in name tag in fasta format (Required)
     :param reference_sequence: Annotated reference sequences in json format (Required)
+    :param threads: Number of threads for multiprocessing (Default: 1)
     :param annotated_json: Output list of sequences annotated in json format (Default: referenced.json)
 
 This file is part of PANGEA HIV project (www.pangea-hiv.org).
 Copyright 2021 Xiaoyu Yu (xiaoyu.yu@ed.ac.uk).
 """
 
 import json
 import sys
 import parasail
-import re
 from Bio import SeqIO
 import datetime as dt
+import multiprocessing
 from genofunc.utils import *
 
-def genome_annotator(raw_fasta,reference_sequence,annotated_json,log_file):
-    def extract_int(base_pos,input_string):
-        temp_string = ""
-        for j in input_string[base_pos:]:
-            if j.isdigit():
-                temp_string += j
-            else:
-                break
-        return int(temp_string)
-    def createCigarDic(cigar):
-        base_position = 1
-        cigar_pos = 0
-        cigar_dic = {}
-        cigar_dic["deletion"] = []
-        cigar_dic["insertion"] = []
-        while cigar_pos < len(cigar):
-            matching_move = extract_int(cigar_pos,cigar)
-            cigar_pos += len(str(matching_move))
-            if cigar[cigar_pos] in ["=","X"]:
-                base_position += matching_move                
-            elif cigar[cigar_pos] == "D":
-                cigar_dic["deletion"].append([base_position,matching_move])
-                base_position += matching_move                
-            elif cigar[cigar_pos] == "I":
-                cigar_dic["insertion"].append([base_position,matching_move])
-                base_position += matching_move                
-            else:
-                print("Invalid Cigar Operator: " + cigar[cigar_pos])
-            cigar_pos += 1
-        return cigar_dic
-
-
-    time_start = dt.datetime.now()
-    location_dic = {}
-    reference_dic = {}
-    sequence_dic = {}
-    features = []
-
-    user_matrix = parasail.matrix_create("ACGTRY?", 5, -4)
-    user_matrix[0,2] = -1   #A-G
-    user_matrix[0,4] = 0    #A-R
-    user_matrix[0,6] = 0    #A-?
-    user_matrix[1,3] = -1   #C-T
-    user_matrix[1,5] = 0    #C-Y
-    user_matrix[1,6] = 0    #C-?
-    user_matrix[2,0] = -1   #G-A
-    user_matrix[2,4] = 0    #G-R
-    user_matrix[2,6] = 0    #G-?
-    user_matrix[3,1] = -1   #T-C
-    user_matrix[3,5] = 0    #T-Y
-    user_matrix[3,6] = 0    #T-?
-    user_matrix[4,0] = 0    #R-A
-    user_matrix[4,2] = 0    #R-G
-    user_matrix[4,6] = 0    #R-?
-    user_matrix[5,1] = 0    #Y-C
-    user_matrix[5,3] = 0    #Y-T
-    user_matrix[5,6] = 0    #Y-?
-    user_matrix[6,0] = 0    #?-A
-    user_matrix[6,1] = 0    #?-C
-    user_matrix[6,2] = 0    #?-G
-    user_matrix[6,3] = 0    #?-T
-    user_matrix[6,4] = 0    #?-R
-    user_matrix[6,5] = 0    #?-Y
-    user_matrix[6,5] = 0    #?-N
-    user_matrix[6,6] = 0    #?-?
+user_matrix = parasail.matrix_create("ACGTRY?", 5, -4)
+user_matrix[0,2] = -1   #A-G
+user_matrix[0,4] = 0    #A-R
+user_matrix[0,6] = 0    #A-?
+user_matrix[1,3] = -1   #C-T
+user_matrix[1,5] = 0    #C-Y
+user_matrix[1,6] = 0    #C-?
+user_matrix[2,0] = -1   #G-A
+user_matrix[2,4] = 0    #G-R
+user_matrix[2,6] = 0    #G-?
+user_matrix[3,1] = -1   #T-C
+user_matrix[3,5] = 0    #T-Y
+user_matrix[3,6] = 0    #T-?
+user_matrix[4,0] = 0    #R-A
+user_matrix[4,2] = 0    #R-G
+user_matrix[4,6] = 0    #R-?
+user_matrix[5,1] = 0    #Y-C
+user_matrix[5,3] = 0    #Y-T
+user_matrix[5,6] = 0    #Y-?
+user_matrix[6,0] = 0    #?-A
+user_matrix[6,1] = 0    #?-C
+user_matrix[6,2] = 0    #?-G
+user_matrix[6,3] = 0    #?-T
+user_matrix[6,4] = 0    #?-R
+user_matrix[6,5] = 0    #?-Y
+user_matrix[6,5] = 0    #?-N
+user_matrix[6,6] = 0    #?-?
+
+time_start = dt.datetime.now()
+location_dic = {}
+reference_dic = {}
+sequence_dic = {}
+out_dic = {}
+features = []
+processing_list = []
+
+def extract_int(base_pos,input_string):
+    temp_string = ""
+    for j in input_string[base_pos:]:
+        if j.isdigit():
+            temp_string += j
+        else:
+            break
+    return int(temp_string)
+
+def createCigarDic(cigar):
+    base_position = 1
+    cigar_pos = 0
+    cigar_dic = {}
+    cigar_dic["deletion"] = []
+    cigar_dic["insertion"] = []
+    while cigar_pos < len(cigar):
+        matching_move = extract_int(cigar_pos,cigar)
+        cigar_pos += len(str(matching_move))
+        if cigar[cigar_pos] in ["=","X"]:
+            base_position += matching_move                
+        elif cigar[cigar_pos] == "D":
+            cigar_dic["deletion"].append([base_position,matching_move])
+            base_position += matching_move                
+        elif cigar[cigar_pos] == "I":
+            cigar_dic["insertion"].append([base_position,matching_move])
+            base_position += matching_move                
+        else:
+            print("Invalid Cigar Operator: " + cigar[cigar_pos])
+        cigar_pos += 1
+    return cigar_dic
+
+def mapping_reference(ref_sequence,query_sequence,id,reference_strain):
+    cigarResults = parasail.sg_trace_striped_32(ref_sequence, query_sequence, 10, 1, user_matrix)
+    reference_dic[id] = {}
+    reference_dic[id]["sequence"] = cigarResults.traceback.ref
+    query_cigar_dic = createCigarDic(cigarResults.cigar.decode.decode('ascii'))
+    for genes in features:
+        if genes not in location_dic[reference_strain].keys():
+            continue
+        else:
+            coordinates = location_dic[reference_strain][genes].split("|")
+            for i in range(len(coordinates)):
+                counter = 0
+                temp_value = int(coordinates[i])
+                gene_coordinate = abs(int(coordinates[i]))
+                for j in query_cigar_dic["deletion"]:
+                    deletion_base = int(j[0])
+                    if deletion_base <= gene_coordinate:
+                        counter += int(j[1])
+                    else:
+                        break
+                coordinates[i] = str(gene_coordinate + counter)
+                if temp_value < 0:
+                    coordinates[i] = str((gene_coordinate + counter)*(-1))
+                else:
+                    coordinates[i] = str(gene_coordinate + counter)
+            reference_dic[id][genes] = "|".join(coordinates)
+    return reference_dic
 
+def genome_annotator(raw_fasta,reference_sequence,threads,annotated_json,log_file):
     if not file_check(raw_fasta):
         print("Input file does not exist. Please check the file path.")
         sys.exit()
     if not file_check(reference_sequence):
         print("Reference file does not exist. Please check the file path.")
         sys.exit()    
 
     log_handle = get_log_handle(log_file)
 
     with open(reference_sequence,"r") as f:
         reference_data = json.load(f)
 
-    for k,v in reference_data.items():
+    for v in reference_data.values():
         for i in v:
             sequence_dic[i["accession"]] = i["sequence"]
             location_dic[i["accession"]] = {}
             location_dic[i["accession"]]["sequence"] = i["sequence"]
             for genes in i["locations"]["location"]:
                 temp_list = []
                 if genes["featureId"] not in features:
                     features.append(genes["featureId"])
                 if isinstance(genes["fragment"], list):
                     for j in genes["fragment"]:
                         temp_list.append(j["from"])
                         temp_list.append(j["to"])
                 else:
-                    for k,v in genes["fragment"].items():
+                    for v in genes["fragment"].values():
                         temp_list.append(v)
                 location_dic[i["accession"]][genes["featureId"]] = "|".join(temp_list)
 
     for record in SeqIO.parse(raw_fasta, "fasta"):
         temp_list = record.id.split("|")
         id = temp_list[0]
         reference_strain = temp_list[-1]
-        cigarResults = parasail.sg_trace_striped_32(str(sequence_dic[reference_strain]), str(record.seq), 10, 1, user_matrix)
-        reference_dic[id] = {}
-        reference_dic[id]["sequence"] = cigarResults.traceback.ref
-        query_cigar_dic = createCigarDic(cigarResults.cigar.decode.decode('ascii'))
-        for genes in features:
-            if genes not in location_dic[reference_strain].keys():
-                log_handle.write(record.id + " does not contain " + genes + " region.\n")
-                continue
+        ref_seq = str(sequence_dic[reference_strain])
+        query_seq = str(record.seq)
+        processing_list.append([ref_seq,query_seq,id,reference_strain])
+    
+    with multiprocessing.Pool(processes=int(threads)) as pool:
+        temp_dic = pool.starmap(mapping_reference, processing_list)
+
+    for i in temp_dic:
+        if type(i) is dict:
+            if len(i) > 0:
+                for j in i:
+                    if j not in out_dic.keys():
+                        out_dic.update(i)
             else:
-                coordinates = location_dic[reference_strain][genes].split("|")
-                for i in range(len(coordinates)):
-                    counter = 0
-                    temp_value = int(coordinates[i])
-                    gene_coordinate = abs(int(coordinates[i]))
-                    for j in query_cigar_dic["deletion"]:
-                        deletion_base = int(j[0])
-                        if deletion_base <= gene_coordinate:
-                            counter += int(j[1])
-                        else:
-                            break
-                    coordinates[i] = str(gene_coordinate + counter)
-                    if temp_value < 0:
-                        coordinates[i] = str((gene_coordinate + counter)*(-1))
-                    else:
-                        coordinates[i] = str(gene_coordinate + counter)
-                reference_dic[id][genes] = "|".join(coordinates)
+                continue
 
     with open(annotated_json, 'w') as f:
-        json_dumps_str = json.dumps(reference_dic, indent=4)
+        json_dumps_str = json.dumps(out_dic, indent=4)
         print(json_dumps_str, file=f)
 
     time_ran = dt.datetime.now() - time_start
     print("Time Lapse:", time_ran.total_seconds() / 60, "Minutes")
 
     close_handle(log_handle)
```

## genofunc/subcommands/genome_annotator.py

```diff
@@ -6,10 +6,11 @@
 from genofunc.genome_annotator import *
 
 def run(options):
 
     genome_annotator(
         options.raw_fasta,
         options.reference_sequence,
+        options.threads,
         options.out_annotation,
         options.log_file
     )
```

## Comparing `genofunc-1.1.1.dist-info/METADATA` & `genofunc-1.1.2.dist-info/METADATA`

 * *Files 18% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.1
 Name: genofunc
-Version: 1.1.1
+Version: 1.1.2
 Summary: UNKNOWN
 Home-page: https://github.com/xiaoyu518/genofunc
 Author: UNKNOWN
 Author-email: UNKNOWN
 License: MIT
 Platform: UNKNOWN
 Classifier: Topic :: Scientific/Engineering :: Bio-Informatics
```

## Comparing `genofunc-1.1.1.dist-info/RECORD` & `genofunc-1.1.2.dist-info/RECORD`

 * *Files 9% similar despite different names*

```diff
@@ -1,34 +1,34 @@
 genofunc/__init__.py,sha256=GTcLsdLpkTw75d8hjasWQOV_sTsM0cmvWRsY9NCOvco,537
-genofunc/__main__.py,sha256=m3vOExraa1TYoRWqfsSUV-pY9EOl7nDtRf1tUyprfuU,16448
+genofunc/__main__.py,sha256=8NZ2-4jkJhUhafT0yqX8WpNF4FwDnbI3oLQRvEROk2s,16649
 genofunc/concatenate_fasta.py,sha256=kZAxSll72Dbcsi-SOfz8WO0Fk6YUncGn9-HZzYEXGm0,1419
 genofunc/extract_metadata.py,sha256=OhdtJbAEzOyXOBaxf8k2ffuDWpeBnSnzkTHLTn4ngH8,2932
 genofunc/feature_extractor.py,sha256=6kotHlO0PxqgV40O_ZI-7wb3UKhPXQ9bQqJb73jyqq8,3781
 genofunc/filter_fasta.py,sha256=d65e2v51m8HRt1MQgfJq03owvCcz2FJ0KDa491AqF2g,4847
 genofunc/gene_concatenator.py,sha256=U0VyT6R4HKRwtH54cb_h8873rlmfXRoriyY_D1dSI3U,1576
-genofunc/genome_annotator.py,sha256=f2lnfLavSFx2dAzeyNimB_q6mhCZynyVtkQX_15Etwg,6117
+genofunc/genome_annotator.py,sha256=XwSlmXoJUprEkYWcpbh-KNRckUTY-92fN2mwUiMubZE,6411
 genofunc/genome_annotator2.py,sha256=4PUGUaZKJYd-DJIPhHcKYvyLVT7fbKMlqAalktKz4tM,7691
 genofunc/group_align.py,sha256=qj0jnLN3v-sE1UjApBDHosGRlfcdBCoNKBrjfcvEErQ,4677
 genofunc/merge.py,sha256=XHb626y6u5plNmcfejRXmgzO3D17pW7J4nC96PnvYUk,4106
 genofunc/name_splitter.py,sha256=LMvmfMCSGSHKrShR-bIim5nALq4Q1dKV30X4xpCnIH4,1587
 genofunc/reference_matcher.py,sha256=FnhCDGcDiFXK5ZGETVw5y3-Dvs4d4z2ZdkUzYSAq9yQ,3924
 genofunc/rename_fasta.py,sha256=RpIto02NkmdxnpcbPL2rlw69MZ6jMm4Lrx4aq3Q08bA,1521
 genofunc/strain_encoder.py,sha256=tCwzxo5V9Y9_JHApxfL0RX8wg0epgsHk4yBDbX_N_Jc,2789
 genofunc/utils.py,sha256=K0KHBLHCTm9Kgv07IpSllSfHyNzjKBOgPIO7ZYdl4Ng,650
 genofunc/subcommands/__init__.py,sha256=ULmKBq-2JqrAJrh0LKy6yDG_7a4GeTmE7sXjPEmQOHM,410
 genofunc/subcommands/concatenate_fasta.py,sha256=3JD1N14vgcbbAWYKRi725x3OwaI7DLUimbd6_WahGoc,350
 genofunc/subcommands/extract_metadata.py,sha256=ITzNfDsG2WyPhSX1P_SGy6Z57tsJLAjg6O9vIuHp7BM,536
 genofunc/subcommands/feature_extractor.py,sha256=NBTwaasbkeKU2bo3wFqENSK_mwuiO5NYuSIKzVAl8h0,418
 genofunc/subcommands/filter_fasta.py,sha256=4Hdj-34OnhliBn43wK9OBXQkHYU2RJfW6tieWhkMwbE,448
 genofunc/subcommands/gene_concatenator.py,sha256=SIt0jOl-cGE5SukmKLlwPm6hVFUpMzLIMPk9ICPIpno,308
-genofunc/subcommands/genome_annotator.py,sha256=VNF7r9-mtnnqxGJBxnImBmUoJK6hmFBrnJzuCzEw3NQ,334
+genofunc/subcommands/genome_annotator.py,sha256=kesTdOZ5l_j2haJdSr4SAbUH1S43k7IozH4h1yQeC6s,359
 genofunc/subcommands/group_align.py,sha256=4dERT7cdM_pgHANrE1WMpBPF03sn11rk0csETmrIU5A,342
 genofunc/subcommands/merge.py,sha256=uTz_oVJ1Tzd958jttdPo8ka3jWIKc0dt4quEuwu2VIw,325
 genofunc/subcommands/name_splitter.py,sha256=s3SMtbRsODSx0RuK6wcByeUJfpPgYLfdMqI1nMJaJYs,373
 genofunc/subcommands/reference_matcher.py,sha256=ArmUvL6JFD8NV2gBMdZrNDnQ4Szby0QLSXKPEQ1oO9w,391
 genofunc/subcommands/rename_fasta.py,sha256=TACml7SbX7ZRRVbJoGNW49I6HdjBxAwj3TVi3N2ClsM,317
 genofunc/subcommands/strain_encoder.py,sha256=wBtrfxFMERiDAEtAziu145M1QChp6prDpGi9RhA5MSk,397
-genofunc-1.1.1.dist-info/METADATA,sha256=dWaTx7q1VeiFog8PbvLvdwM5HqiYAggU7NaI_h2zXjc,572
-genofunc-1.1.1.dist-info/WHEEL,sha256=U88EhGIw8Sj2_phqajeu_EAi3RAo8-C6zV3REsWbWbs,92
-genofunc-1.1.1.dist-info/entry_points.txt,sha256=9xT6cQgMyyC1ktKvIEtsvI1NFjO_W9iloZ9NbIcPw6Q,53
-genofunc-1.1.1.dist-info/top_level.txt,sha256=rdj71U7IfsnpHsF93VlZG0FTDrPHDxaua4sx_F5yYUI,9
-genofunc-1.1.1.dist-info/RECORD,,
+genofunc-1.1.2.dist-info/METADATA,sha256=Y8uv0_z7gTeZgg8_GTWSNHVsBomRdbLuqMi4wQfunJM,572
+genofunc-1.1.2.dist-info/WHEEL,sha256=U88EhGIw8Sj2_phqajeu_EAi3RAo8-C6zV3REsWbWbs,92
+genofunc-1.1.2.dist-info/entry_points.txt,sha256=9xT6cQgMyyC1ktKvIEtsvI1NFjO_W9iloZ9NbIcPw6Q,53
+genofunc-1.1.2.dist-info/top_level.txt,sha256=rdj71U7IfsnpHsF93VlZG0FTDrPHDxaua4sx_F5yYUI,9
+genofunc-1.1.2.dist-info/RECORD,,
```

